### DISCUSSION TITLE
How to make a custom cursor invert its color relative to the background?

### DISCUSSION THREAD

**Discussion by evhenii-komar** [AvaloniaUI/Avalonia#10947] [Q&A]
Greetings!
How can I make the custom cursor invert the color of its pixels relative to the background color as the **IBeam** text cursor does? The cur file of cursors with this behavior has an icon with zero alpha channel in Windows and works fine in WPF. As I understood Avalonia doesn't support **cur** format so I use a **png** file. I tried to use it in the same way and just made visible pixels of the cursor white with a transparent alpha channel. However, it renders as just a gray square. If leave alpha channel 255 for the white-colored pixels the cursor looks fine but it is always white, with no color inversion in this case. So it isn't visible on the white background of the application window. I need it to work as a regular IBeam text cursor which always has a color inverted to the background. Is it possible to make it in Avalonia for the custom cursors?

Here is how it looks with the **png** icon in Avalonia:
![image](https://user-images.githubusercontent.com/85240195/230664710-5c400a86-8537-4bca-bede-962364ddacf7.png)

Here is how it looks with the **cur** icon in WPF:
![image](https://user-images.githubusercontent.com/85240195/230665183-71f402a9-9c99-4d16-9161-34625b58677c.png)
![image](https://user-images.githubusercontent.com/85240195/230665613-a790cf8f-bb9c-4fa6-b0f9-52cb1bc08603.png)

Here is **cur** icon and its **png** equivalent:
[cursor_icons.zip](https://github.com/AvaloniaUI/Avalonia/files/11181064/cursor_icons.zip)
If you'll try to see the png icon in the image editor you will not see anything because of the alpha channel. The alpha channel is zero. It is the same as in the cur icon that works in WPF. You can invert the alpha channel of the picture to be sure it isn't empty.

I'll be glad for any help.

**Comment by sb-viktor** [comment]
I need the same thing with text (with a picture on the back for example).

**Reply by thevortexcloud** [reply]
If you are generating the text as part of the picture (EG via Skia), you can sample the pixels behind the text. But it's a bit complicated to do.

**Reply by sb-viktor** [reply]
Wow, I didn't think of that. It really could probably be done, but it doesn't really work well for me. I'd like it to just be a **property** that can be changed to `true` so that the text is converted to negative and can be seen on both black and white for example

**Reply by thevortexcloud** [reply]
You could also do a similar thing with a converter to work out the average colour in a region of the image. It likely however wont be fast or play well with layouts (since the text/image will likely move around a lot). If you really wanted a property, you could also probably write your own attached property. But again, it's complicated.

https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-attached-properties

As a guess, you may also need to access Avalonia's Skia Canvas directly (which you can do via some not well documented methods) to work out what is even behind the text. But this is not something I have ever tried.

In any case, this is a somewhat niche feature and I doubt it will be added to Avalonia itself. Most people are using static elements for this sort of thing/not putting text directly over a picture. EG you could much more easily work around this by just putting a semi transparent background behind the text (such as ~50% black).
